typedb/concept/type/thing_type.py

```diff
@@ -18,103 +18,166 @@
-from typedb.common.rpc.request_builder import thing_type_set_supertype_req, thing_type_get_instances_req, \
-    thing_type_set_abstract_req, thing_type_unset_abstract_req, thing_type_set_plays_req, thing_type_set_owns_req, \
-    thing_type_get_plays_req, thing_type_get_owns_req, thing_type_unset_plays_req, thing_type_unset_owns_req, \
-    thing_type_get_owns_explicit_req, thing_type_get_owns_overridden_req, thing_type_get_plays_explicit_req, \
-    thing_type_get_plays_overridden, thing_type_get_syntax_req
-from typedb.concept.proto import concept_proto_builder, concept_proto_reader
-from typedb.concept.type.type import _Type, _RemoteType
-
-
-class _ThingType(ThingType, _Type):
-
-    def as_remote(self, transaction):
-        return _RemoteThingType(transaction, self.get_label(), self.is_root(), self.is_abstract())
-
-    def as_thing_type(self) -> "ThingType":
+from __future__ import annotations
+from abc import abstractmethod, ABC
+from itertools import chain
+from typing import Optional, Iterator, TYPE_CHECKING, Any
+
+from typedb.api.concept.type.thing_type import ThingType
+from typedb.common.label import Label
+from typedb.common.streamer import Streamer
+from typedb.common.transitivity import Transitivity
+from typedb.concept.concept_factory import attribute_type_of, role_type_of
+from typedb.concept.type.type import _Type
+from typedb.typedb_client_python import thing_type_is_root, thing_type_is_abstract, \
+    thing_type_get_label, thing_type_delete, thing_type_is_deleted, thing_type_set_label, thing_type_set_abstract, \
+    thing_type_unset_abstract, thing_type_set_plays, thing_type_unset_plays, thing_type_set_owns, thing_type_get_owns, \
+    thing_type_get_plays, thing_type_get_owns_overridden, thing_type_unset_owns, thing_type_get_syntax, \
+    thing_type_get_plays_overridden, concept_iterator_next
+
+if TYPE_CHECKING:
+    from typedb.api.concept.type.attribute_type import AttributeType
+    from typedb.api.concept.value.value import ValueType
+    from typedb.api.concept.type.annotation import Annotation
+    from typedb.concept.type.attribute_type import _AttributeType
+    from typedb.concept.type.role_type import _RoleType
+    from typedb.concept.thing.thing import _Thing
+    from typedb.connection.transaction import _Transaction
+
+
+class _ThingType(ThingType, _Type, ABC):
+
+    def as_thing_type(self) -> ThingType:
         return self
 
+    def is_root(self) -> bool:
+        return thing_type_is_root(self.native_object)
+
+    def is_abstract(self) -> bool:
+        return thing_type_is_abstract(self.native_object)
 
-class _RemoteThingType(_RemoteType, RemoteThingType):
+    def get_label(self) -> Label:
+        return Label.of(thing_type_get_label(self.native_object))
 
-    def as_remote(self, transaction):
-        return _RemoteThingType(transaction, self.get_label(), self.is_root(), self.is_abstract())
+    def delete(self, transaction: _Transaction) -> None:
```

Comment:
What happens if `delete()` is called, and then the user calls a method on the Python instance which already has been deleted in the Rust layer, so the native object is 'invalid' (if that's how we've designed it)

Reply:
In progress

Reply:
We should get an exception from the server. Now the error is `The transaction is closed because of the error(s):
status: Internal, message: "", details: [], metadata: MetadataMap { headers: {} }`, and it has to be fixed at the server side.